File: mobile/iot/AndroidScooterBleBridge.java
package com.example.taurus;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothProfile;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanResult;
import android.bluetooth.le.ScanSettings;
import android.content.Context;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.os.ParcelUuid;
import android.util.SparseArray;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;

import org.json.JSONException;
import org.json.JSONObject;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

public class AndroidScooterBleBridge {
    private static final UUID SERVICE_UUID = UUID.fromString("7b6a1000-2f6d-4e7f-9b2e-30a0bbd80101");
    private static final UUID COMMAND_UUID = UUID.fromString("7b6a1001-2f6d-4e7f-9b2e-30a0bbd80101");
    private static final UUID TELEMETRY_UUID = UUID.fromString("7b6a1002-2f6d-4e7f-9b2e-30a0bbd80101");
    private static final UUID CCCD_UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");
    private static final int NEARBY_MANUFACTURER_ID = 0xFFFF;
    private static final int NEARBY_PROTOCOL_VERSION = 1;

    private final Context context;
    private final WebView webView;
    private final BluetoothAdapter bluetoothAdapter;

    private BluetoothGatt gatt;
    private BluetoothGattCharacteristic commandCharacteristic;
    private BluetoothGattCharacteristic telemetryCharacteristic;
    private String pendingScooterId = "";
    private String pendingCommand = "";
    private String lastWrittenCommand = "";
    private String lastTelemetryJson = "";
    private ScanCallback activeScanCallback;
    private ScanCallback nearbyScanCallback;
    private boolean nearbyScanRequested = false;
    private boolean connecting = false;
    private int nearbyPacketCount = 0;
    private int nearbyScooterCount = 0;
    private long lastNearbyDebugMs = 0;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    public AndroidScooterBleBridge(Context context, WebView webView, BluetoothAdapter bluetoothAdapter) {
        this.context = context;
        this.webView = webView;
        this.bluetoothAdapter = bluetoothAdapter;
    }

    @JavascriptInterface
    public String connect(String json) {
        try {
            JSONObject payload = new JSONObject(json);
            String scooterId = payload.optString("scooterId").trim();
            String mac = payload.optString("mac").trim();

            if (scooterId.equals(pendingScooterId) && commandCharacteristic != null && gatt != null) {
                emitStatus("Already connected to " + scooterId);
                emitConnected(scooterId);
                return "{\"ok\":true,\"alreadyConnected\":true}";
            }

            if (scooterId.equals(pendingScooterId) && connecting) {
                emitStatus("Bluetooth connection already in progress.");
                return "{\"ok\":true,\"connecting\":true}";
            }

            pendingScooterId = scooterId;

            if (!mac.isEmpty()) {
                connectByMac(mac);
                return "{\"ok\":true,\"mac\":true}";
            }

            scanForScooter(pendingScooterId);
            return "{\"ok\":true}";
        } catch (JSONException error) {
            emitStatus("Invalid Bluetooth request.");
            return "{\"ok\":false,\"error\":\"invalid_json\"}";
        }
    }

    @JavascriptInterface
    public String sendCommand(String json) {
        try {
            JSONObject payload = new JSONObject(json);
            String command = payload.optString("command").trim();

            if (commandCharacteristic == null || gatt == null) {
                pendingCommand = command;
                emitStatus("Command queued until Bluetooth connects.");
                return "{\"ok\":true,\"queued\":true}";
            }

            writeCommand(command);
            return "{\"ok\":true}";
        } catch (JSONException error) {
            return "{\"ok\":false,\"error\":\"invalid_json\"}";
        }
    }

    @JavascriptInterface
    @SuppressLint("MissingPermission")
    public String readTelemetry(String json) {
        if (gatt == null || telemetryCharacteristic == null) {
            return "{\"ok\":false,\"error\":\"not_connected\"}";
        }

        boolean started = gatt.readCharacteristic(telemetryCharacteristic);

        if (!started) {
            emitStatus("Telemetry read could not start.");
        }

        return "{\"ok\":" + (started ? "true" : "false") + ",\"telemetry\":" + (lastTelemetryJson.isEmpty() ? "null" : lastTelemetryJson) + "}";
    }

    @JavascriptInterface
    public String getLastTelemetry() {
        return lastTelemetryJson.isEmpty() ? "{}" : lastTelemetryJson;
    }

    @JavascriptInterface
    @SuppressLint("MissingPermission")
    public String disconnect(String json) {
        if (gatt != null) {
            gatt.disconnect();
            gatt.close();
            gatt = null;
        }

        commandCharacteristic = null;
        telemetryCharacteristic = null;
        connecting = false;
        emitStatus("Bluetooth disconnected after command.");
        resumeNearbyScan();

        return "{\"ok\":true}";
    }

    @JavascriptInterface
    @SuppressLint("MissingPermission")
    public String startNearbyScan() {
        nearbyScanRequested = true;

        if (bluetoothAdapter == null || bluetoothAdapter.getBluetoothLeScanner() == null) {
            emitNearbyScanStatus("Bluetooth scanner unavailable.");
            return "{\"ok\":false,\"error\":\"scanner_unavailable\"}";
        }

        if (nearbyScanCallback != null) {
            emitNearbyScanStatus(
                    "BLE scanner active: " + nearbyPacketCount
                            + " packets, " + nearbyScooterCount + " scooters."
            );
            return "{\"ok\":true,\"alreadyScanning\":true}";
        }

        ScanSettings settings = new ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_BALANCED)
                .build();

        nearbyScanCallback = new ScanCallback() {
            @Override
            public void onScanResult(int callbackType, ScanResult result) {
                handleNearbyResult(result);
            }

            @Override
            public void onBatchScanResults(List<ScanResult> results) {
                for (ScanResult result : results) {
                    handleNearbyResult(result);
                }
            }

            @Override
            public void onScanFailed(int errorCode) {
                nearbyScanCallback = null;
                emitNearbyScanStatus("Nearby scooter scan failed: " + errorCode);
            }
        };

        nearbyPacketCount = 0;
        nearbyScooterCount = 0;
        lastNearbyDebugMs = 0;
        bluetoothAdapter.getBluetoothLeScanner().startScan(null, settings, nearbyScanCallback);
        emitNearbyScanStatus("BLE scanner started. Waiting for advertisements...");
        return "{\"ok\":true}";
    }

    @SuppressLint("MissingPermission")
    private void pauseNearbyScan() {
        if (nearbyScanCallback == null || bluetoothAdapter == null
                || bluetoothAdapter.getBluetoothLeScanner() == null) {
            return;
        }

        bluetoothAdapter.getBluetoothLeScanner().stopScan(nearbyScanCallback);
        nearbyScanCallback = null;
    }

    private void resumeNearbyScan() {
        if (nearbyScanRequested) {
            startNearbyScan();
        }
    }

    @SuppressLint("MissingPermission")
    private void connectByMac(String mac) {
        if (bluetoothAdapter == null) {
            emitStatus("Bluetooth adapter unavailable.");
            return;
        }

        try {
            pauseNearbyScan();

            if (gatt != null) {
                gatt.disconnect();
                gatt.close();
                gatt = null;
                commandCharacteristic = null;
            }

            connecting = true;
            BluetoothDevice device = bluetoothAdapter.getRemoteDevice(mac);
            emitStatus("Connecting by MAC " + mac);
            connectGatt(device);
            scheduleConnectTimeout();
        } catch (IllegalArgumentException error) {
            connecting = false;
            emitStatus("Invalid scooter MAC: " + mac);
        }
    }

    @SuppressLint("MissingPermission")
    private void scanForScooter(String scooterId) {
        if (bluetoothAdapter == null || bluetoothAdapter.getBluetoothLeScanner() == null) {
            emitStatus("Bluetooth scanner unavailable.");
            return;
        }

        if (scooterId == null || scooterId.trim().isEmpty()) {
            emitStatus("Scan scooter IoT QR first.");
            return;
        }

        pauseNearbyScan();

        if (gatt != null) {
            gatt.disconnect();
            gatt.close();
            gatt = null;
            commandCharacteristic = null;
        }

        String deviceName = "RYDOZ-" + scooterId;

        ScanSettings settings = new ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
                .build();

        emitStatus("Scanning for " + deviceName);
        connecting = true;

        activeScanCallback = new ScanCallback() {
            @Override
            public void onScanResult(int callbackType, ScanResult result) {
                handleScanResult(result, deviceName);
            }

            @Override
            public void onBatchScanResults(List<ScanResult> results) {
                for (ScanResult result : results) {
                    handleScanResult(result, deviceName);
                }
            }

            @Override
            public void onScanFailed(int errorCode) {
                connecting = false;
                emitStatus("BLE scan failed: " + errorCode);
                resumeNearbyScan();
            }
        };

        bluetoothAdapter.getBluetoothLeScanner().startScan(null, settings, activeScanCallback);

        mainHandler.postDelayed(() -> {
            if (activeScanCallback != null && commandCharacteristic == null) {
                bluetoothAdapter.getBluetoothLeScanner().stopScan(activeScanCallback);
                activeScanCallback = null;
                connecting = false;
                emitStatus("Scooter not found. Keep ESP32 on and close nRF Connect.");
                resumeNearbyScan();
            }
        }, 12000);
    }

    @SuppressLint("MissingPermission")
    private void handleScanResult(ScanResult result, String expectedName) {
        if (activeScanCallback == null || result == null) {
            return;
        }

        String deviceName = result.getDevice().getName();
        boolean hasScooterService = false;

        if ((deviceName == null || deviceName.isEmpty()) && result.getScanRecord() != null) {
            deviceName = result.getScanRecord().getDeviceName();
        }

        if (result.getScanRecord() != null && result.getScanRecord().getServiceUuids() != null) {
            hasScooterService = result.getScanRecord().getServiceUuids().contains(new ParcelUuid(SERVICE_UUID));
        }

        if ((deviceName == null || !deviceName.equals(expectedName)) && !hasScooterService) {
            return;
        }

        bluetoothAdapter.getBluetoothLeScanner().stopScan(activeScanCallback);
        activeScanCallback = null;
        emitStatus("Connecting to " + (deviceName != null ? deviceName : "scooter service"));
        connectGatt(result.getDevice());
        scheduleConnectTimeout();
    }

    private void handleNearbyResult(ScanResult result) {
        if (result == null || result.getScanRecord() == null) {
            return;
        }

        nearbyPacketCount++;
        SparseArray<byte[]> manufacturerData = result.getScanRecord().getManufacturerSpecificData();
        byte[] payload = manufacturerData == null
                ? null
                : manufacturerData.get(NEARBY_MANUFACTURER_ID);

        if (payload == null || payload.length < 3 || (payload[0] & 0xFF) != NEARBY_PROTOCOL_VERSION) {
            long now = System.currentTimeMillis();

            if (now - lastNearbyDebugMs >= 3000) {
                lastNearbyDebugMs = now;
                emitNearbyScanStatus(
                        "BLE scanner active: " + nearbyPacketCount
                                + " packets seen, no RYDOZ battery packet."
                );
            }
            return;
        }

        int battery = payload[1] & 0xFF;
        String scooterId = new String(
                Arrays.copyOfRange(payload, 2, payload.length),
                StandardCharsets.UTF_8
        ).trim();

        if (scooterId.isEmpty() || battery > 100) {
            emitNearbyScanStatus("RYDOZ packet found but its battery data is invalid.");
            return;
        }

        nearbyScooterCount++;
        emitNearbyScanStatus("Found " + scooterId + " at " + battery + "%.");
        emitNearbyScooter(scooterId, battery, result.getRssi(), result.getDevice().getAddress());
    }

    @SuppressLint("MissingPermission")
    private void connectGatt(BluetoothDevice device) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            gatt = device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE);
        } else {
            gatt = device.connectGatt(context, false, gattCallback);
        }
    }

    private void scheduleConnectTimeout() {
        mainHandler.postDelayed(() -> {
            if (connecting && commandCharacteristic == null) {
                connecting = false;
                if (gatt != null) {
                    gatt.disconnect();
                    gatt.close();
                    gatt = null;
                }
                emitStatus("Bluetooth connect timeout. Restart ESP32 and try again.");
                resumeNearbyScan();
            }
        }, 12000);
    }

    private final BluetoothGattCallback gattCallback = new BluetoothGattCallback() {
        @SuppressLint("MissingPermission")
        @Override
        public void onConnectionStateChange(BluetoothGatt bluetoothGatt, int status, int newState) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                connecting = true;
                bluetoothGatt.discoverServices();
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                commandCharacteristic = null;
                connecting = false;
                if (status != BluetoothGatt.GATT_SUCCESS) {
                    bluetoothGatt.close();
                    if (gatt == bluetoothGatt) {
                        gatt = null;
                    }
                }
                resumeNearbyScan();
            }
        }

        @SuppressLint("MissingPermission")
        @Override
        public void onServicesDiscovered(BluetoothGatt bluetoothGatt, int status) {
            if (status != BluetoothGatt.GATT_SUCCESS) {
                connecting = false;
                emitStatus("Service discovery failed: " + status);
                return;
            }

            if (bluetoothGatt.getService(SERVICE_UUID) == null) {
                connecting = false;
                emitStatus("Scooter BLE service not found.");
                return;
            }

            commandCharacteristic = bluetoothGatt.getService(SERVICE_UUID).getCharacteristic(COMMAND_UUID);
            telemetryCharacteristic = bluetoothGatt.getService(SERVICE_UUID).getCharacteristic(TELEMETRY_UUID);

            if (telemetryCharacteristic != null) {
                bluetoothGatt.setCharacteristicNotification(telemetryCharacteristic, true);
                BluetoothGattDescriptor descriptor = telemetryCharacteristic.getDescriptor(CCCD_UUID);

                if (descriptor != null) {
                    descriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
                    bluetoothGatt.writeDescriptor(descriptor);
                }
            }

            connecting = false;
            emitConnected(pendingScooterId);

            if (pendingCommand != null && !pendingCommand.isEmpty()) {
                writeCommand(pendingCommand);
                pendingCommand = "";
            }
        }

        @Override
        public void onCharacteristicChanged(BluetoothGatt bluetoothGatt, BluetoothGattCharacteristic characteristic) {
            if (TELEMETRY_UUID.equals(characteristic.getUuid())) {
                String telemetry = new String(characteristic.getValue(), StandardCharsets.UTF_8);
                handleTelemetry(telemetry);
            }
        }

        @Override
        public void onCharacteristicRead(BluetoothGatt bluetoothGatt, BluetoothGattCharacteristic characteristic, int status) {
            if (TELEMETRY_UUID.equals(characteristic.getUuid()) && status == BluetoothGatt.GATT_SUCCESS) {
                String telemetry = new String(characteristic.getValue(), StandardCharsets.UTF_8);
                handleTelemetry(telemetry);
            }
        }

        @Override
        public void onCharacteristicWrite(BluetoothGatt bluetoothGatt, BluetoothGattCharacteristic characteristic, int status) {
            if (COMMAND_UUID.equals(characteristic.getUuid())) {
                if (status != BluetoothGatt.GATT_SUCCESS) {
                    emitStatus("Command write failed: " + status);
                }
            }
        }
    };

    private void emitStatus(String message) {
        emitJs("window.dispatchEvent(new CustomEvent('scooter:ble-status',{detail:{message:" + JSONObject.quote(message) + "}}))");
    }

    @SuppressLint("MissingPermission")
    private void writeCommand(String command) {
        if (commandCharacteristic == null || gatt == null) {
            pendingCommand = command;
            return;
        }

        commandCharacteristic.setWriteType(BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT);
        commandCharacteristic.setValue((command + "\n").getBytes(StandardCharsets.UTF_8));
        lastWrittenCommand = command;

        boolean started = gatt.writeCharacteristic(commandCharacteristic);

        if (!started) {
            emitStatus("Command write could not start: " + command);
        }
    }

    private void emitConnected(String scooterId) {
        emitJs("window.dispatchEvent(new CustomEvent('scooter:ble-connected',{detail:{scooterId:" + JSONObject.quote(scooterId) + "}}))");
    }

    private void handleTelemetry(String json) {
        lastTelemetryJson = json;
        emitTelemetry(json);
    }

    private void emitTelemetry(String json) {
        emitJs("window.dispatchEvent(new CustomEvent('scooter:telemetry',{detail:" + json + "}))");
    }

    private void emitNearbyScooter(String scooterId, int battery, int rssi, String mac) {
        emitJs(
                "window.dispatchEvent(new CustomEvent('scooter:nearby-telemetry',{detail:{"
                        + "scooterId:" + JSONObject.quote(scooterId)
                        + ",battery:" + battery
                        + ",rssi:" + rssi
                        + ",mac:" + JSONObject.quote(mac)
                        + "}}))"
        );
    }

    private void emitNearbyScanStatus(String message) {
        emitJs(
                "window.dispatchEvent(new CustomEvent('scooter:nearby-scan-status',{detail:{message:"
                        + JSONObject.quote(message)
                        + "}}))"
        );
    }

    private void emitJs(String script) {
        webView.post(() -> webView.evaluateJavascript(script, null));
    }
}

File: mobile/iot/AndroidScooterBleBridge.kt
package com.rydoz.app.iot

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Build
import android.os.ParcelUuid
import android.webkit.JavascriptInterface
import android.webkit.WebView
import org.json.JSONObject
import java.util.UUID

class AndroidScooterBleBridge(
    private val context: Context,
    private val webView: WebView,
    private val bluetoothAdapter: BluetoothAdapter
) {
    private val serviceUuid = UUID.fromString("7b6a1000-2f6d-4e7f-9b2e-30a0bbd80101")
    private val commandUuid = UUID.fromString("7b6a1001-2f6d-4e7f-9b2e-30a0bbd80101")
    private val telemetryUuid = UUID.fromString("7b6a1002-2f6d-4e7f-9b2e-30a0bbd80101")
    private val cccdUuid = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

    private var gatt: BluetoothGatt? = null
    private var commandCharacteristic: BluetoothGattCharacteristic? = null
    private var telemetryCharacteristic: BluetoothGattCharacteristic? = null
    private var latestTelemetryJson: String? = null
    private var pendingScooterId: String = ""
    private var nearbyScanCallback: ScanCallback? = null

    @JavascriptInterface
    fun connect(json: String): String {
        val payload = JSONObject(json)
        pendingScooterId = payload.optString("scooterId").trim()
        disconnectGatt()
        scanForScooter(pendingScooterId)
        return """{"ok":true}"""
    }

    @JavascriptInterface
    fun sendCommand(json: String): String {
        val command = JSONObject(json).optString("command").trim()
        val characteristic = commandCharacteristic ?: return """{"ok":false,"error":"not_connected"}"""
        characteristic.value = "$command\n".toByteArray(Charsets.UTF_8)
        gatt?.writeCharacteristic(characteristic)
        return """{"ok":true}"""
    }

    @JavascriptInterface
    fun readTelemetry(json: String = "{}"): String {
        telemetryCharacteristic?.let { characteristic ->
            gatt?.readCharacteristic(characteristic)
        }

        val telemetry = latestTelemetryJson
        return if (telemetry.isNullOrBlank()) {
            """{"ok":true,"telemetry":null}"""
        } else {
            """{"ok":true,"telemetry":$telemetry}"""
        }
    }

    @JavascriptInterface
    fun getLastTelemetry(): String {
        return latestTelemetryJson ?: """{"raw":null}"""
    }

    @JavascriptInterface
    fun disconnect(json: String = "{}"): String {
        disconnectGatt()
        emitStatus("Bluetooth disconnected")
        return """{"ok":true}"""
    }

    @JavascriptInterface
    fun startNearbyScan(json: String = "{}"): String {
        scanNearbyScooters()
        return """{"ok":true}"""
    }

    @SuppressLint("MissingPermission")
    private fun scanForScooter(scooterId: String) {
        val scanner = bluetoothAdapter.bluetoothLeScanner ?: return emitStatus("Bluetooth scanner unavailable")
        val filters = listOf(
            ScanFilter.Builder()
                .setDeviceName("RYDOZ-$scooterId")
                .setServiceUuid(ParcelUuid(serviceUuid))
                .build()
        )
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        emitStatus("Scanning for RYDOZ-$scooterId")
        scanner.startScan(filters, settings, object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                scanner.stopScan(this)
                emitStatus("Connecting to ${result.device.name ?: scooterId}")
                gatt = result.device.connectGatt(context, false, gattCallback)
            }

            override fun onScanFailed(errorCode: Int) {
                emitStatus("BLE scan failed: $errorCode")
            }
        })
    }

    private val gattCallback = object : BluetoothGattCallback() {
        @SuppressLint("MissingPermission")
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                gatt.discoverServices()
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                commandCharacteristic = null
                telemetryCharacteristic = null
                emitStatus("Bluetooth disconnected")
            }
        }

        @SuppressLint("MissingPermission")
        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            val service = gatt.getService(serviceUuid) ?: return emitStatus("Scooter service not found")
            commandCharacteristic = service.getCharacteristic(commandUuid)
            val telemetry = service.getCharacteristic(telemetryUuid)
            telemetryCharacteristic = telemetry

            if (telemetry != null) {
                gatt.setCharacteristicNotification(telemetry, true)
                telemetry.getDescriptor(cccdUuid)?.let { descriptor ->
                    descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    gatt.writeDescriptor(descriptor)
                }
            }

            emitConnected(pendingScooterId)
            telemetry?.let { gatt.readCharacteristic(it) }
        }

        override fun onCharacteristicRead(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            status: Int
        ) {
            if (characteristic.uuid == telemetryUuid && status == BluetoothGatt.GATT_SUCCESS) {
                emitTelemetry(characteristic.value.toString(Charsets.UTF_8))
            }
        }

        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
            if (characteristic.uuid == telemetryUuid) {
                emitTelemetry(characteristic.value.toString(Charsets.UTF_8))
            }
        }
    }

    @SuppressLint("MissingPermission")
    private fun scanNearbyScooters() {
        val scanner = bluetoothAdapter.bluetoothLeScanner ?: return emitNearbyScanStatus("Bluetooth scanner unavailable")
        nearbyScanCallback?.let { scanner.stopScan(it) }

        val filters = listOf(
            ScanFilter.Builder()
                .setServiceUuid(ParcelUuid(serviceUuid))
                .build()
        )
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        nearbyScanCallback = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                emitNearbyTelemetry(result)
            }

            override fun onBatchScanResults(results: MutableList<ScanResult>) {
                results.forEach { emitNearbyTelemetry(it) }
            }

            override fun onScanFailed(errorCode: Int) {
                emitNearbyScanStatus("Nearby scan failed: $errorCode")
            }
        }

        emitNearbyScanStatus("Searching for powered scooters...")
        scanner.startScan(filters, settings, nearbyScanCallback)
    }

    @SuppressLint("MissingPermission")
    private fun emitNearbyTelemetry(result: ScanResult) {
        val deviceName = result.device.name ?: return
        if (!deviceName.startsWith("RYDOZ-")) {
            return
        }

        val scooterId = deviceName.removePrefix("RYDOZ-")
        val telemetry = parseTelemetry(result.scanRecord?.serviceData?.values?.firstOrNull())
        val battery = telemetry?.optDouble("battery", Double.NaN)

        if (battery == null || battery.isNaN()) {
            return
        }

        val json = JSONObject()
            .put("scooterId", scooterId)
            .put("battery", battery.toInt())
            .put("rssi", result.rssi)
            .put("mac", result.device.address)

        emitJs("window.dispatchEvent(new CustomEvent('scooter:nearby-telemetry',{detail:$json}))")
    }

    private fun parseTelemetry(bytes: ByteArray?): JSONObject? {
        if (bytes == null || bytes.isEmpty()) {
            return null
        }

        return try {
            JSONObject(bytes.toString(Charsets.UTF_8))
        } catch (_: Exception) {
            null
        }
    }

    @SuppressLint("MissingPermission")
    private fun disconnectGatt() {
        commandCharacteristic = null
        telemetryCharacteristic = null
        gatt?.disconnect()
        gatt?.close()
        gatt = null
    }

    private fun emitStatus(message: String) {
        emitJs("window.dispatchEvent(new CustomEvent('scooter:ble-status',{detail:{message:${JSONObject.quote(message)}}}))")
    }

    private fun emitNearbyScanStatus(message: String) {
        emitJs("window.dispatchEvent(new CustomEvent('scooter:nearby-scan-status',{detail:{message:${JSONObject.quote(message)}}}))")
    }

    private fun emitConnected(scooterId: String) {
        emitJs("window.dispatchEvent(new CustomEvent('scooter:ble-connected',{detail:{scooterId:${JSONObject.quote(scooterId)}}}))")
    }

    private fun emitTelemetry(json: String) {
        latestTelemetryJson = json
        emitJs("window.dispatchEvent(new CustomEvent('scooter:telemetry',{detail:$json}))")
    }

    private fun emitJs(script: String) {
        webView.post { webView.evaluateJavascript(script, null) }
    }
}

/*
MainActivity setup notes:

webView.settings.javaScriptEnabled = true
webView.addJavascriptInterface(
    AndroidScooterBleBridge(this, webView, BluetoothAdapter.getDefaultAdapter()),
    "ScooterBle"
)

AndroidManifest permissions:
<uses-permission android:name="android.permission.BLUETOOTH_SCAN" />
<uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
<uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" android:maxSdkVersion="30" />

At runtime request BLUETOOTH_SCAN and BLUETOOTH_CONNECT on Android 12+.
Request ACCESS_FINE_LOCATION on Android 11 and lower for BLE scanning.
*/
